server/src/main/java/org/elasticsearch/cluster/routing/BatchedRerouteService.java

```diff
@@ -114,22 +142,26 @@ public void onFailure(String source, Exception e) {
                             logger.error(() -> new ParameterizedMessage("unexpected failure during [{}], current state version [{}]",
                                 source, state.version()), e);
                         }
-                        currentListeners.onFailure(new ElasticsearchException("delayed reroute [" + reason + "] failed", e));
+                        currentListeners.forEach(l ->
+                            l.onFailure(new ElasticsearchException("delayed reroute [" + reason + "] failed", e)));
                     }
 
                     @Override
                     public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
-                        currentListeners.onResponse(null);
+                        currentListeners.forEach(l -> l.onResponse(null));
```

Comment:
Same issue as with the failures here as well. We probably should use `org.elasticsearch.action.ActionListener#onResponse(java.lang.Iterable<org.elasticsearch.action.ActionListener<Response>>, Response)` to not fail the loop on the first exception? (Also, this makes the behavior equivalent to what it was where `PlainListenableActionFuture` would do the same and ensure all the listeners are called).